dev_setup.sh

```diff
@@ -0,0 +1,142 @@
+# added to this script.
+#
+
+set -e
+
+HERE=$(cd `dirname $0`; pwd)
+LOG=$HERE/dev_setup.log
+ROOT=$HERE
+ENV=$ROOT/env.sh
+
+WHAT4_SOLVERS_SNAPSHOT="snapshot-20240212"
+WHAT4_SOLVERS_URL="https://github.com/GaloisInc/what4-solvers/releases/download/$WHAT4_SOLVERS_SNAPSHOT/"
+WHAT4_SOLVERS_MACOS_12="macos-12-X64-bin.zip"
+WHAT4_SOLVERS_MACOS_14="macos-14-ARM64-bin.zip"
+
+function notice {
+    echo "[NOTICE] $*"
+}
+
+# Requires: LOG set to log file path.
+function logged {
+    if ! [ -z "$LOG" ]
+    then
+        mkdir -p `dirname $LOG`
+        echo "$@" >>$LOG
+        if ! "$@" >>$LOG 2>&1
+        then
+            echo
+            echo "An error occurred; please see $LOG"
+            echo "Here are the last 50 lines:"
+            tail -n 50 $LOG
+            exit 1
+        fi
+    fi
+}
+
+function update_submodules {
+    cd $ROOT
+    notice "Updating submodules"
+    logged git submodule update --init
+}
+
+function install_ghcup {
+    if ! ghcup --version &> /dev/null
+    then
+        notice "Installing ghcup, GHC, and cabal"
+        # Technically the installation only requires cabal, but it's
+        # recommended to get the whole GCH shebang in one package.
+        # The output is not routed to log because the installer is interactive.
+        curl --proto '=https' --tlsv1.2 -sSf https://get-ghcup.haskell.org | sh
+    else
+        notice "Using existing ghcup installation"
+    fi
+
+}
+
+function is_macos {
+    [ "$OSTYPE" = "darwin"* ]
+}
+
+# Indicate whether this is running macOS on the Apple M series hardware.
+function is_macos_aarch {
+    # Is it running macOS?
+    if [ ! is_macos ]; then
+        return 1
+    fi
+
+    # Does it use an M-series chip?
+    # Actually this command apparently doesn't exist in macOS 12, so this will fail the wrong way
+    chip=$(system_profiler SPHardwareDataType | grep Chip)
+    [ $chip == *M1* || $chip == *M2* || $chip == *M3* ]
+}
+
+function install_gmp {
+    if [ is_macos ]; then
+        notice "Installing GMP via Homebrew, if it's not already installed"
+        logged brew install gmp
+
+        # `brew --prefix` is different on macOS 12 and macOS 14
+        echo "export CPATH=$(brew --prefix)/include" >> $ENV
+        echo "export LIBRARY_PATH=$(brew --prefix)/lib" >> $ENV
+        notice "You may need to source new environment variables added here:\n" \
+            "\$ source $ENV"
+    else
+        notice "Did not install GMP. This script only supports macOS 12 and 14"
+    fi
+}
+
+# Installs the two solvers required to run the test suite for the repo.
+# Users may want to install other solvers, and indeed the what4 solvers repo
+# includes a half dozen other solvers that are compatible with cryptol. 
+function install_what4_solvers {
+    if ! cvc4 --version &> /dev/null || ! cvc5 --version &> /dev/null; then
+        notice "Installing cvc4 and/or cvc5 solvers"
+
+        if [ ! is_macos ]; then
```

Comment:
This OS check is done inside this function. Is there a reason it can't be done at the top level? I find it odd for it to be here because it halts the setup process, and that's the sort of check I tend to assume would be done before any real work begins, so that once work begins, assumptions like "we're on a supported platform" can be thought to hold for the rest of the script.

Reply:
I think the idea is to put something like
```
case $(uname -s) of
    Darwin) CRYPTOL_PLATFORM=macos;;
    Linux) if [ -f /etc/redhat_release ]; then CRYPTOL_PLATFORM=redhat else ...
    *) echo "Unsupported platform" 2>&1; exit 1;;
esac
```
at the top and then use the results of that probe in whatever other conditional logic is required later.

Reply:
This is much nicer, thanks.